From: https://patch-diff.githubusercontent.com/raw/MariaDB/server/pull/2896.patch
Author: Tuukka Pasanen <tuukka.pasanen@ilmi.fi>
Date: Mon, 4 Dec 2023 10:24:07 +0200
Subject: [PATCH] MDEV-32936: Detect do ARM processor have NEON intrinsics or
 not

NEON is available from ARMv7 and up. Debian has armel toolchain
which is targeted to ARMv5/6 processors. If processor does not have
NEON then it should be emulated. Commit adds detection do processor
have NEON or not and add correct building flags if it does not
have them.
---
 mysys/CMakeLists.txt | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

--- a/mysys/CMakeLists.txt
+++ b/mysys/CMakeLists.txt
@@ -89,6 +89,12 @@ ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "a
     include(CheckCXXSourceCompiles)
 
     CHECK_CXX_SOURCE_COMPILES("
+    // print error 'NEON intrinsics not available with the soft-float ABI.
+    // Please use -mfloat-abi=softfp or -mfloat-abi=hard'
+    #include <arm_neon.h>
+    int main(){}" HAVE_ARMV8_NEON)
+
+    CHECK_CXX_SOURCE_COMPILES("
     #define CRC32CX(crc, value) __asm__(\"crc32cx %w[c], %w[c], %x[v]\":[c]\"+r\"(crc):[v]\"r\"(value))
     asm(\".arch_extension crc\");
     unsigned int foo(unsigned int ret) {
@@ -134,9 +140,23 @@ ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "a
       IF(HAVE_ARMV8_CRYPTO)
         ADD_DEFINITIONS(-DHAVE_ARMV8_CRYPTO)
       ENDIF()
+      IF(NOT HAVE_ARMV8_NEON)
+        set(CMAKE_REQUIRED_FLAGS "-mfpu=neon -mfloat-abi=softfp")
+
+        CHECK_CXX_SOURCE_COMPILES("
+        #include <arm_neon.h>
+        int main(){}" HAVE_ARMV8_NEON_SOFTFP)
+      ENDIF()
       SET(MYSYS_SOURCES ${MYSYS_SOURCES} crc32/crc32_arm64.c)
+
+      IF(HAVE_ARMV8_NEON_SOFTFP)
+        SET(CRC32_ARM64_FLAGS "-mfpu=neon -mfloat-abi=softfp -march=armv8-a+crc+crypto")
+      ELSE()
+        SET(CRC32_ARM64_FLAGS "-march=armv8-a+crc+crypto")
+      ENDIF()
+
       SET_SOURCE_FILES_PROPERTIES(crc32/crc32_arm64.c PROPERTIES
-        COMPILE_FLAGS "-march=armv8-a+crc+crypto")
+        COMPILE_FLAGS ${CRC32_ARM64_FLAGS})
     ENDIF()
   ENDIF()
 ENDIF()
